src/utils/schemaTree.js

```diff
@@ -0,0 +1,312 @@
+import { clone, assocPath } from 'ramda';
+import { COMBINATION_TYPES, COMPLEX_TYPES } from './constants';
+
+/**
+ * Generate a tree that illustrates the recursive structure of a schema.
+ * A single node within the tree defines the following structure:
+ * {
+ *   schema: ..       // the schema the node is representing
+ *   children: [..]   // node objects of the same structure (for subschemas)
+ *   path: [..]       // the path from root to node, array of child indexes in
+ *                       sequential order (for $refs and indentation level)
+ *                       ex. [1, 0] = node at root.children[1].children[0]
+ *                                    also has depth of 1 for indent size.
+ * }

```

Comment:
Ended up creating a tree structure rather than using a copy of the schema object! 

I initially thought that since the schemaTree would pretty much be the same structure as the schema itself, it would suffice to just use a deep copy of the schema as the schemaTree. However, I quickly noticed we would need to manage `path` and `isExpanded` states for refNodes and it would be hard to maintain them with just using the schema object copy 😅 .

Reply:
> Hope it's okay I changed to creating a tree structure instead

That's totally fine!